cpp/ql/src/experimental/Security/CWE/CWE-783/OperatorPrecedenceLogicErrorWhenUseBitwiseOrLogicalOperations.ql

```diff
@@ -0,0 +1,179 @@
+/**
+ * @name Operator Precedence Logic Error When Use Bitwise Or Logical Operations
+ * @description --Finding places to use bit and logical operations, without explicit priority allocation.
+ *              --For example, `a || b ^ c` and `(a || b) ^ c` give different results when `b` is zero.
+ * @kind problem
+ * @id cpp/operator-precedence-logic-error-when-use-bitwise-logical-operations
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-783
+ *       external/cwe/cwe-480
+ */
+
+import cpp
+import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+
+/** Holds if `exptmp` equals expression logical or followed by logical and. */
+predicate isLogicalOrAndExpr(Expr exptmp) {
+  not exptmp.(LogicalOrExpr).getLeftOperand() instanceof BinaryOperation and
+  not exptmp.(LogicalOrExpr).getRightOperand().isParenthesised() and
+  exptmp.(LogicalOrExpr).getRightOperand() instanceof LogicalAndExpr
+}
+
+/** Holds if `exptmp` equals expression logical or followed by bit operation. */
+predicate isLogicalOrandBitwise(Expr exptmp) {
+  not exptmp.(LogicalOrExpr).getLeftOperand() instanceof BinaryOperation and
+  not exptmp.(LogicalOrExpr).getRightOperand().isParenthesised() and
+  (
+    exptmp.(LogicalOrExpr).getRightOperand().(BinaryBitwiseOperation).getLeftOperand().getType()
+      instanceof BoolType and
+    not exptmp
+        .(LogicalOrExpr)
+        .getRightOperand()
+        .(BinaryBitwiseOperation)
+        .getRightOperand()
+        .getValue() = "0" and
+    not exptmp
+        .(LogicalOrExpr)
+        .getRightOperand()
+        .(BinaryBitwiseOperation)
+        .getRightOperand()
+        .getValue() = "1"
+  )
+  or
+  not exptmp.(LogicalAndExpr).getLeftOperand() instanceof BinaryOperation and
+  not exptmp.(LogicalAndExpr).getRightOperand().isParenthesised() and
+  (
+    exptmp.(LogicalAndExpr).getRightOperand().(BinaryBitwiseOperation).getLeftOperand().getType()
+      instanceof BoolType and
+    not exptmp
+        .(LogicalAndExpr)
+        .getRightOperand()
+        .(BinaryBitwiseOperation)
+        .getRightOperand()
+        .getValue() = "0" and
+    not exptmp
+        .(LogicalAndExpr)
+        .getRightOperand()
+        .(BinaryBitwiseOperation)
+        .getRightOperand()
+        .getValue() = "1"
```

Comment:
the idea of this and the previous section (32-43) is to find a situation in which the right-hand operand of a bit operation is not limited to 0 or 1.
in this case, the logical operation will be performed with the result of a binary operation that is not a Boolean type.
in my opinion this indicates a priority error. after all, it will be quite difficult for a developer to evaluate the conversion of the results of a bit operation to a boolean type.

Reply:
I agree with those comments. Do you mind writing this down in the QL file as well (so we have that for future maintainability)?